core/src/test/scala/unit/kafka/zk/ZooKeeperTestHarness.scala

```diff
@@ -54,3 +58,40 @@ abstract class ZooKeeperTestHarness extends JUnitSuite with Logging {
     Configuration.setConfiguration(null)
   }
 }
+
+object ZooKeeperTestHarness {
+
+  /**
+   * Verify that a previous test that doesn't use ZooKeeperTestHarness hasn't left behind an unexpected thread
+   */
+  @BeforeClass
+  def setUpClass() {
+    verifyNoUnexpectedThreads()
+  }
+
+  /**
+   * Verify that tests from the current test class using ZooKeeperTestHarness hasn't left behind an unexpected thread
+   */
+  @AfterClass
+  def tearDownClass() {
+    verifyNoUnexpectedThreads()
+  }
+
+  /**
+   * Verifies that threads which are known to cause transient failures in subsequent tests
+   * have been shutdown. These include threads which make connections to brokers and may
+   * cause issues when broker ports are reused (e.g. auto-create topics) as well as threads
+   * which reset static JAAS configuration.
+   */
+  def verifyNoUnexpectedThreads() {
+    val unexpectedThreadNames = Set("controller-event-thread",
+                                    "kafka-producer-network-thread",
+                                    "kafka-coordinator-heartbeat-thread",
+                                    "ZkClient-EventThread")
+    def allThreads = Thread.getAllStackTraces.keySet.asScala.map(thread => thread.getName())
+    val (threads, noUnexpected) = TestUtils.computeUntilTrue(allThreads) { threads =>
+      threads.forall(t => unexpectedThreadNames.forall(s => !t.contains(s)))
+    }
+    assertTrue(s"Unexpected threads $threads", noUnexpected)
```

Comment:
Are you intending to output all threads, or just the expected ones? From the message, it seems like it's the latter, but all are being printed. A simple change to make it print the unexpected ones would be to change `allThreads` to `unexpectedThreads`. The predicate would then just check that this was empty.